test/DeviceRunners.Cli.Tests/NetworkServiceTests.cs

```diff
@@ -1,11 +1,72 @@
 using DeviceRunners.Cli.Services;
 using System.Net.Sockets;
 using System.Text;
+using System.Net;
 
 namespace DeviceRunners.Cli.Tests;
 
 public class NetworkServiceTests
 {
+    private class NetworkEventLog
+    {
+        public List<string> Events { get; } = new();
+        public List<string> DataReceived { get; } = new();
+        
+        public void LogConnectionEstablished(object? sender, ConnectionEventArgs e)
+        {
+            Events.Add($"ConnectionEstablished:{e.RemoteEndPoint}");
+        }
+        
+        public void LogConnectionClosed(object? sender, ConnectionEventArgs e)
+        {
+            Events.Add($"ConnectionClosed:{e.RemoteEndPoint}");
+        }
+        
+        public void LogDataReceived(object? sender, DataReceivedEventArgs e)
+        {
+            Events.Add($"DataReceived:{e.RemoteEndPoint}:{e.Data}");
+            DataReceived.Add(e.Data);
+        }
+        
+        public bool HasConnectionEstablished => Events.Any(e => e.StartsWith("ConnectionEstablished:"));
+        public bool HasConnectionClosed => Events.Any(e => e.StartsWith("ConnectionClosed:"));
+        public bool HasDataReceived => Events.Any(e => e.StartsWith("DataReceived:"));
+    }
+    
+    private static NetworkService CreateServiceWithEventLog(out NetworkEventLog eventLog)
```

Comment:
This should rather accept the service and then setup sw the test owns the service:

```suggestion
    private static NetworkEventLog CreateEventLog(NetworkService service)
```